### ISSUE TITLE
KL Loss

### ISSUE THREAD

**Issue by CDitzel** [openai/DALL-E#26]
I am having trouble getting the dVAE to train properly if I include the KL loss term with a uniform prior over the number of visual tokens. Does anyone here has had similar experiences or problems? The paper mentions an increasing schedule for the kl weight factor but I cant get it to work properly and results are always better if I set the KL loss to zero altogether.

Maybe someone can help?

**Comment by adityaramesh** [comment]
I'm assuming you got this to work but let me know if you still have issues.

**Comment by CDitzel** [comment]
I guess I am missing something. What exactly do you use to parameterize the posterior to the uniform prior? the logits of the encoder? or do you you some form of distance metric to the codebook vectors? the codebook itself, as I take it from your code is the first conv layer of the decoder?

is the overall loss then scaled like

(beta * KL /192 + Reconstruction_loss) / (3*256*256)>

i.e. the ratio of the number of pixels of the decoder output to the spatial extend of the encoder output ?

thank you for your help

**Comment by adityaramesh** [comment]
> What exactly do you use to parameterize the posterior to the uniform prior? the logits of the encoder?

Yes, I use the logits of the encoder for the KL term and pretend that they correspond to a categorical distribution for the purpose of computing KL to the uniform categorical prior, even when the relaxation temperature is greater than zero. I think this is mentioned in the appendix of the paper.

> is the overall loss then scaled like

I think the normalization factor for the KL term should be beta / (32 * 32), and the normalization factor for the reconstruction loss should be (256 * 256 * 3).

**Comment by CDitzel** [comment]
> > What exactly do you use to parameterize the posterior to the uniform prior? the logits of the encoder?
> 
> Yes, I use the logits of the encoder for the KL term and pretend that they correspond to a categorical distribution for the purpose of computing KL to the uniform categorical prior, even when the relaxation temperature is greater than zero. I think this is mentioned in the appendix of the paper.
> 

thank you Aditya for getting back to me on this one. Much appreciated!

Using the encoder output as parameteres for the gumbel softmax does not work, not for Andrej Karpathy, nor for me or others in the community so far. What does work is to use the eucledian distance of the 32x32 field logits to the "codebook" vector and use those distances as paramterizaton for the gumbel softmax and its distribution in the KL loss. but it seems, this is more than you guys did and results are still not very good. 

> > is the overall loss then scaled like
> 
> I think the normalization factor for the KL term should be beta / (32 * 32), and the normalization factor for the reconstruction loss should be (256 * 256 * 3).

so the (256 * 256 * 3) division only happens for the reconstruction loss and not the sum of both terms then? in the paper it said, that the overall loss is scaled with this factor
